Make error clearer when a binary fails to load (iltorb) because it was compiled for a different version/paltform

The original report is kept at the end of this issue and it has evolved into the following:

sonarwhal uses `iltorb` which is a binary package. When updating versions of node without recompiling it, sonarwhal will fail loading saying that it can't find the resource `jsdom`. We should capture better this error for future users, node updates frequently so this will most likely happen again.

```
$ sonarwhal https://example.com
Error: Resource jsdom not found
    at Object.exports.loadResource (/usr/local/lib/node_modules/sonarwhal/dist/src/lib/utils/resource-loader.js:172:15)
    at Object.exports.loadResources (/usr/local/lib/node_modules/sonarwhal/dist/src/lib/utils/resource-loader.js:213:29)
    at exports.analyze (/usr/local/lib/node_modules/sonarwhal/dist/src/lib/cli/analyze.js:112:38)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
    at Function.Module.runMain (module.js:678:11)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3
? There is 1 package from your .sonarwhalrc file not installed or with an incompatible version. Do you want us to try to install/update them? Yes
Installing packages...
✖ Error executing "npm install @sonarwhal/[object Object] -g"
Try executing:
    sudo npm install @sonarwhal/[object Object] -g
            manually to install all the packages.
```

-----------------------------------

Hi there.
I try to use sonarwhal but execution failed.

I think it is wired: `Can't require @sonarwhal/configuration-progressive-web-apps` (see output section)
@sonarwhal/configuration-progressive-web-apps exists into node_modules.

```
ls -la node_modules/@sonarwhal/configuration-progressive-web-apps/
total 56
drwxr-xr-x   7 leko  staff    238 May  6 17:49 .
drwxr-xr-x  11 leko  staff    374 May  6 17:49 ..
-rw-r--r--   1 leko  staff    305 Oct 26  1985 CHANGELOG.md
-rw-r--r--   1 leko  staff  11384 Oct 26  1985 LICENSE.txt
-rw-r--r--   1 leko  staff   1590 Oct 26  1985 README.md
-rw-r--r--   1 leko  staff    444 Oct 26  1985 index.json
-rw-r--r--   1 leko  staff   2236 May  6 17:49 package.json
```

## Environment

<!-- You get the version numbers for the following by running:

 * sonarwhal -v
 * node -v
 * npm -v

-->

* __`sonarwhal` version:v1.8.1__
* __`Node.js` version:v8.9.0 and v10.0.0__
* __`npm` version:5.8.0 and 5.6.0__

I tested two version of Node.js (8.9.0 and 10.0.0)

## `sonarwhal` configuration

<details>
<summary>sonarwhal’s configuration</summary>

<!--
Note: The `sonarwhal` configuration may be in a file named `.sonarwhalrc`,
      or in  the `package.json` file under the `sonarwhalConfig` propery.

Paste those configurations below:
-->

```js
{
    "extends": [
        "progressive-web-apps"
    ]
}
```

</details>

## URL for which `sonarwhal` failed

* __URL:https://example.com__

## Output

<details>
<summary>sonarwhal’s raw output</summary>

<!-- Paste the raw output below: -->

```text
npx sonarwhal --debug https://example.com
  sonarwhal:utils:misc Loading JSON file: /usr/local/lib/node_modules/sonarwhal/package.json +0ms
  sonarwhal:utils:get-as-uri Adding valid target: https://example.com/ +0ms
  sonarwhal:cli:analyze Loading configuration file from /Users/leko/.ghq/github.com/Leko/ansitsu/.sonarwhalrc. +0ms
  sonarwhal:utils:misc Loading JSON file: /Users/leko/.ghq/github.com/Leko/ansitsu/.sonarwhalrc +97ms
  sonarwhal:config:config-validator Validating configuration +0ms
  sonarwhal:config Composing configuration from extends +0ms
  sonarwhal:utils:resource-loader Searching progressive-web-apps… +0ms
  sonarwhal:utils:resource-loader Can't require @sonarwhal/configuration-progressive-web-apps +1ms
  sonarwhal:utils:resource-loader Can't require sonarwhal-configuration-progressive-web-apps +0ms
  sonarwhal:utils:resource-loader Can't require /usr/local/lib/node_modules/sonarwhal/dist/src/lib/configurations/progressive-web-apps/progressive-web-apps.js +1ms
  sonarwhal:utils:resource-loader Can't require /Users/leko/.ghq/github.com/Leko/ansitsu +2ms
  sonarwhal:utils:resource-loader Resource progressive-web-apps not found +0ms
Error: Resource progressive-web-apps not found
    at Object.exports.loadResource (/usr/local/lib/node_modules/sonarwhal/dist/src/lib/utils/resource-loader.js:172:15)
    at Object.exports.loadConfiguration (/usr/local/lib/node_modules/sonarwhal/dist/src/lib/utils/resource-loader.js:208:20)
    at userConfig.extends.map (/usr/local/lib/node_modules/sonarwhal/dist/src/lib/config.js:38:52)
    at Array.map (<anonymous>)
    at composeConfig (/usr/local/lib/node_modules/sonarwhal/dist/src/lib/config.js:37:47)
    at Function.fromConfig (/usr/local/lib/node_modules/sonarwhal/dist/src/lib/config.js:161:28)
    at exports.analyze (/usr/local/lib/node_modules/sonarwhal/dist/src/lib/cli/analyze.js:111:45)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
    at Function.Module.runMain (module.js:678:11)
    at findNodeScript.then.existing (/usr/local/lib/node_modules/npm/node_modules/libnpx/index.js:267:14)
    at <anonymous>
Exit code: 1
```

</details>

@molant I got it. output is here:

```
sonarwhal --debug  https://example.com
  sonarwhal:utils:misc Loading JSON file: /usr/local/lib/node_modules/sonarwhal/package.json +0ms
  sonarwhal:utils:get-as-uri Adding valid target: https://example.com/ +0ms
  sonarwhal:cli:analyze Loading configuration file from /Users/leko/Downloads/aaa/.sonarwhalrc. +0ms
  sonarwhal:utils:misc Loading JSON file: /Users/leko/Downloads/aaa/.sonarwhalrc +53ms
  sonarwhal:config:config-validator Validating configuration +0ms
  sonarwhal:config Composing configuration from extends +0ms
  sonarwhal:utils:resource-loader Searching progressive-web-apps… +0ms
  sonarwhal:utils:resource-loader progressive-web-apps found in @sonarwhal/configuration-progressive-web-apps +1ms
  sonarwhal:config:config-validator Validating configuration +15ms
  sonarwhal:config Composing configuration from extends +1ms
  sonarwhal:config overriding config settings with values provided via CLI +1ms
  sonarwhal:config Initializing ignored urls +3ms
  sonarwhal:utils:resource-loader Searching jsdom… +4ms
  sonarwhal:utils:resource-loader Package sonarwhal-configuration-progressive-web-apps not found +40ms
  sonarwhal:utils:resource-loader Can't require @sonarwhal/connector-jsdom +0ms
  sonarwhal:utils:resource-loader Can't require sonarwhal-connector-jsdom +0ms
  sonarwhal:utils:resource-loader Can't require /usr/local/lib/node_modules/sonarwhal/dist/src/lib/connectors/jsdom/jsdom.js +361ms
  sonarwhal:utils:resource-loader Can't require /Users/leko/Downloads/aaa +0ms
  sonarwhal:utils:resource-loader Resource jsdom not found +0ms
Error: Resource jsdom not found
    at Object.exports.loadResource (/usr/local/lib/node_modules/sonarwhal/dist/src/lib/utils/resource-loader.js:172:15)
    at Object.exports.loadResources (/usr/local/lib/node_modules/sonarwhal/dist/src/lib/utils/resource-loader.js:213:29)
    at exports.analyze (/usr/local/lib/node_modules/sonarwhal/dist/src/lib/cli/analyze.js:112:38)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
    at Function.Module.runMain (module.js:678:11)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3
  sonarwhal:utils:resource-loader Searching apple-touch-icons… +2ms
  sonarwhal:utils:resource-loader Package sonarwhal-configuration-progressive-web-apps not found +8ms
  sonarwhal:utils:resource-loader Can't require @sonarwhal/rule-apple-touch-icons +0ms
  sonarwhal:utils:resource-loader Can't require sonarwhal-rule-apple-touch-icons +1ms
  sonarwhal:utils:resource-loader Can't require /usr/local/lib/node_modules/sonarwhal/dist/src/lib/rules/apple-touch-icons/apple-touch-icons.js +0ms
  sonarwhal:utils:resource-loader Can't require /Users/leko/Downloads/aaa +0ms
  sonarwhal:utils:resource-loader apple-touch-icons found in /usr/local/lib/node_modules/@sonarwhal/configuration-progressive-web-apps/node_modules/@sonarwhal/rule-apple-touch-icons +4ms
  sonarwhal:utils:resource-loader Searching manifest-app-name… +3ms
  sonarwhal:utils:resource-loader Package sonarwhal-configuration-progressive-web-apps not found +8ms
  sonarwhal:utils:resource-loader Can't require @sonarwhal/rule-manifest-app-name +1ms
  sonarwhal:utils:resource-loader Can't require sonarwhal-rule-manifest-app-name +0ms
  sonarwhal:utils:resource-loader Can't require /usr/local/lib/node_modules/sonarwhal/dist/src/lib/rules/manifest-app-name/manifest-app-name.js +0ms
  sonarwhal:utils:resource-loader Can't require /Users/leko/Downloads/aaa +0ms
  sonarwhal:utils:resource-loader manifest-app-name found in /usr/local/lib/node_modules/@sonarwhal/configuration-progressive-web-apps/node_modules/@sonarwhal/rule-manifest-app-name +2ms
  sonarwhal:utils:resource-loader Searching manifest-exists… +0ms
  sonarwhal:utils:resource-loader Package sonarwhal-configuration-progressive-web-apps not found +9ms
  sonarwhal:utils:resource-loader Can't require @sonarwhal/rule-manifest-exists +0ms
  sonarwhal:utils:resource-loader Can't require sonarwhal-rule-manifest-exists +0ms
  sonarwhal:utils:resource-loader Can't require /usr/local/lib/node_modules/sonarwhal/dist/src/lib/rules/manifest-exists/manifest-exists.js +0ms
  sonarwhal:utils:resource-loader Can't require /Users/leko/Downloads/aaa +0ms
  sonarwhal:utils:resource-loader manifest-exists found in /usr/local/lib/node_modules/@sonarwhal/configuration-progressive-web-apps/node_modules/@sonarwhal/rule-manifest-exists +1ms
  sonarwhal:utils:resource-loader Searching manifest-file-extension… +18ms
  sonarwhal:utils:resource-loader Package sonarwhal-configuration-progressive-web-apps not found +6ms
  sonarwhal:utils:resource-loader Can't require @sonarwhal/rule-manifest-file-extension +1ms
  sonarwhal:utils:resource-loader Can't require sonarwhal-rule-manifest-file-extension +0ms
  sonarwhal:utils:resource-loader Can't require /usr/local/lib/node_modules/sonarwhal/dist/src/lib/rules/manifest-file-extension/manifest-file-extension.js +0ms
  sonarwhal:utils:resource-loader Can't require /Users/leko/Downloads/aaa +0ms
  sonarwhal:utils:resource-loader manifest-file-extension found in /usr/local/lib/node_modules/@sonarwhal/configuration-progressive-web-apps/node_modules/@sonarwhal/rule-manifest-file-extension +1ms
  sonarwhal:utils:resource-loader Searching manifest-is-valid… +0ms
  sonarwhal:utils:resource-loader Package sonarwhal-configuration-progressive-web-apps not found +7ms
  sonarwhal:utils:resource-loader Can't require @sonarwhal/rule-manifest-is-valid +0ms
  sonarwhal:utils:resource-loader Can't require sonarwhal-rule-manifest-is-valid +1ms
  sonarwhal:utils:resource-loader Can't require /usr/local/lib/node_modules/sonarwhal/dist/src/lib/rules/manifest-is-valid/manifest-is-valid.js +0ms
  sonarwhal:utils:resource-loader Can't require /Users/leko/Downloads/aaa +0ms
  sonarwhal:utils:resource-loader manifest-is-valid found in /usr/local/lib/node_modules/@sonarwhal/configuration-progressive-web-apps/node_modules/@sonarwhal/rule-manifest-is-valid +6ms
  sonarwhal:utils:resource-loader Searching manifest… +1ms
  sonarwhal:utils:resource-loader Package sonarwhal-configuration-progressive-web-apps not found +11ms
  sonarwhal:utils:resource-loader Can't require @sonarwhal/parser-manifest +0ms
  sonarwhal:utils:resource-loader Can't require sonarwhal-parser-manifest +0ms
  sonarwhal:utils:resource-loader Can't require /usr/local/lib/node_modules/sonarwhal/dist/src/lib/parsers/manifest/manifest.js +0ms
  sonarwhal:utils:resource-loader Can't require /Users/leko/Downloads/aaa +0ms
  sonarwhal:utils:resource-loader manifest found in /usr/local/lib/node_modules/@sonarwhal/configuration-progressive-web-apps/node_modules/@sonarwhal/parser-manifest +1ms
  sonarwhal:utils:resource-loader Searching summary… +0ms
  sonarwhal:utils:resource-loader Package sonarwhal-configuration-progressive-web-apps not found +7ms
  sonarwhal:utils:resource-loader Can't require @sonarwhal/formatter-summary +0ms
  sonarwhal:utils:resource-loader Can't require sonarwhal-formatter-summary +1ms
  sonarwhal:utils:resource-loader Can't require /usr/local/lib/node_modules/sonarwhal/dist/src/lib/formatters/summary/summary.js +0ms
  sonarwhal:utils:resource-loader Can't require /Users/leko/Downloads/aaa +0ms
  sonarwhal:utils:resource-loader summary found in /usr/local/lib/node_modules/@sonarwhal/configuration-progressive-web-apps/node_modules/@sonarwhal/formatter-summary +21ms
? There is 1 package from your .sonarwhalrc file not installed or with an incompatible version. Do you want us to try to install/update them? Yes
  sonarwhal:utils:npm Running command npm install @sonarwhal/[object Object] -g +0ms
Installing packages...
- Running command: npm install @sonarwhal/[object Object] -g  sonarwhal:utils:npm npm ERR! code EINVALIDTAGNAME
  sonarwhal:utils:npm npm ERR! Invalid tag name "Object]": Tags may not have any characters that encodeURIComponent encodes.
  sonarwhal:utils:npm
  sonarwhal:utils:npm npm ERR! A complete log of this run can be found in:
  sonarwhal:utils:npm npm ERR!     /Users/leko/.npm/_logs/2018-05-08T13_29_29_637Z-debug.log
  sonarwhal:utils:npm  +3s
✖ Error executing "npm install @sonarwhal/[object Object] -g"
Try executing:
    sudo npm install @sonarwhal/[object Object] -g
            manually to install all the packages.
Exit code: 1
```

I debugged about this line.

`  sonarwhal:utils:resource-loader Can't require /usr/local/lib/node_modules/sonarwhal/dist/src/lib/connectors/jsdom/jsdom.js +361ms`

An error is here. 

```
Error: The module '/usr/local/lib/node_modules/sonarwhal/node_modules/iltorb/build/bindings/iltorb.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 64. This version of Node.js requires
NODE_MODULE_VERSION 57. Please try re-compiling or re-installing
...
```

It's my mistake.
I removed sonarwhal globally and re-run `npm install -g --engine-strict sonarwhal`.
It works fine.

```
sonarwhal:utils:resource-loader jsdom found in /usr/local/lib/node_modules/sonarwhal/dist/src/lib/connectors/jsdom/jsdom.js +442ms
...
apple-touch-icons  1 error
manifest-exists    1 error
✖ Found a total of 2 errors and 0 warnings
  sonarwhal:cli:analyze Total runtime: 2564ms +3s
Exit code: 1
```

I installed it into another Node.js version (In v10.0) but I run it in v8.9.
Thanks.

--- a/packages/sonarwhal/src/lib/utils/resource-loader.ts
+++ b/packages/sonarwhal/src/lib/utils/resource-loader.ts
@@ -147,9 +147,17 @@ export const tryToLoadFrom = (resourcePath: string): any => {
         debug(`Can't require ${resourcePath}`);
 
         if (e.code === 'MODULE_NOT_FOUND') {
+            /*
+             * This get the name of the missed module
+             * e.g: Cannot find module 'iltorb'
+             */
             const exec = moduleNameRegex.exec(e.message);
             const moduleName = exec ? exec[1] : null;
 
+            /*
+             * If the module not found is the same as the module
+             * we are trying to load, then is ok.
+             */
             if (!moduleName || moduleName === resourcePath) {
                 return null;
             }

--- a/packages/sonarwhal/src/lib/utils/resource-loader.ts
+++ b/packages/sonarwhal/src/lib/utils/resource-loader.ts
@@ -25,6 +25,7 @@ import { ResourceType } from '../enums/resourcetype';
 const debug: debug.IDebugger = d(__filename);
 const SONARWHAL_ROOT: string = findPackageRoot();
 const NODE_MODULES_ROOT: string = findNodeModulesRoot();
+const moduleNameRegex: RegExp = /[^']*'([^']*)'/g;
 
 /** Cache of resource builders, indexex by resource Id. */
 const resources: Map<string, Resource> = new Map<string, Resource>();
@@ -144,6 +145,20 @@ export const tryToLoadFrom = (resourcePath: string): any => {
         builder = resource.default || resource;
     } catch (e) {
         debug(`Can't require ${resourcePath}`);
+
+        if (e.code === 'MODULE_NOT_FOUND') {
+            const exec = moduleNameRegex.exec(e.message);
+            const moduleName = exec ? exec[1] : null;
+
+            if (!moduleName || moduleName === resourcePath) {
+                return null;
+            }
+
+            // The resourcePath and the module not found are different.
+            throw new Error(`Module ${moduleName} not found when loading ${resourcePath}`);
+        }
+
+        throw e;
     }
 
     return builder;

--- a/packages/sonarwhal/tests/lib/utils/resource-loader.ts
+++ b/packages/sonarwhal/tests/lib/utils/resource-loader.ts
@@ -19,6 +19,44 @@ const cleanCache = () => {
     delete require.cache[cacheKey];
 };
 
+test.serial('tryToLoadFrom throws an error if a dependency is missing', (t) => {
+    const Module = require('module');
+    const resourceLoader = require('../../../src/lib/utils/resource-loader');
+
+    const sandbox = sinon.createSandbox();
+
+    sandbox.stub(Module.prototype, 'require').throws({
+        code: 'MODULE_NOT_FOUND',
+        message: `Cannot load module 'iltorb'`
+    });
+
+    const { message } = t.throws(() => {
+        resourceLoader.tryToLoadFrom('sonarwhal');
+    });
+
+    t.is(message, 'Module iltorb not found when loading sonarwhal');
+
+    sandbox.restore();
+});
+
+test.serial('tryToLoadFrom does nothing if the package itself is missing', (t) => {
+    const Module = require('module');
+    const resourceLoader = require('../../../src/lib/utils/resource-loader');
+
+    const sandbox = sinon.createSandbox();
+
+    sandbox.stub(Module.prototype, 'require').throws({
+        code: 'MODULE_NOT_FOUND',
+        message: `Cannot load module 'sonarwhal'`
+    });
+
+    const resource = resourceLoader.tryToLoadFrom('sonarwhal');
+
+    t.is(resource, null);
+
+    sandbox.restore();
+});
+
 // TODO: Add tests to verify the order of loading is the right one: core -> scoped -> prefixed. This only checks core resources
 test('loadResource looks for resources in the right order (core > @sonarwhal > sonarwhal- ', (t) => {
     cleanCache();

